### DISCUSSION TITLE
Generate syntax highlighting files

### DISCUSSION THREAD

**Discussion by vanillajonathan** [pyparsing/pyparsing#466] [Ideas]
After you create a parser then you have write syntax highlighting files to syntax highlight the language your parser parses. When doing changes to the parser you manually have to update these files.

It would be good if you could generate such syntax highlight files for:
- [Ace](https://github.com/ajaxorg/ace) JavaScript editor.
- [Lezer](https://lezer.codemirror.net/) for the [CodeMirror](https://codemirror.net/) JavaScript editor.
- [Monaco](https://microsoft.github.io/monaco-editor/), Microsoft's JavaScript editor used by VS Code.
- [GtkSourceView](https://gitlab.gnome.org/GNOME/gtksourceview/), the GTK source code editor component used by GNOME Text Editor, GNOME Builder and others.
- [KDE Syntax Highlighting Engine](https://invent.kde.org/frameworks/syntax-highlighting) used by Kate, KDevelop and others. ([API](https://api.kde.org/frameworks/syntax-highlighting/html/index.html))
- [highlight.js](https://highlightjs.org/), syntax highlighting for the web.
- [Sublime Text](https://www.sublimetext.com/docs/syntax.html), text editor
- [TextMate](https://macromates.com/manual/en/language_grammars), text editor

**Comment by ptmcg** [comment]
That's an interesting idea. What does such a file look like?

My initial thought is that this will be a fairly large add-on to pyparsing, and probably deserving of being its own separate project.

**Reply by vanillajonathan** [reply]
Here are examples:
- [Python in Ace](https://github.com/ajaxorg/ace/blob/master/src/mode/python_highlight_rules.js)
- [Python in Lezer](https://github.com/lezer-parser/python/tree/main/src) (for CodeMirror)
- [Python in Monaco](https://github.com/microsoft/monaco-editor/tree/main/src/basic-languages/python) (used by VS Code)
- [Python in GtkSourceView](https://gitlab.gnome.org/GNOME/gtksourceview/-/blob/master/data/language-specs/python3.lang)
- [Python in KDE Syntax Highlighting Engine](https://invent.kde.org/frameworks/syntax-highlighting/-/blob/master/data/syntax/python.xml)
- [Python in highlight.js](https://github.com/highlightjs/highlight.js/blob/main/src/languages/python.js)

**Reply by ptmcg** [reply]
The Python in Monaco link gives me a 404 error.

**Reply by vanillajonathan** [reply]
My bad, I edited the post and fixed it.

**Comment by ptmcg** [comment]
Is there a pyparsing parser that you have hand-written the syntax highlighting for? Some of your examples seem to follow a common style of Javascript - is there a standard for this?

**Reply by vanillajonathan** [reply]
No I have not, and no there is no real standard; some are in JavaScript, others such as GtkSourceView and KDE Syntax Highlighting Engine are in XML.

**Comment by ptmcg** [comment]
I agree, this sounds like an interesting project, but I have no bandwidth to give it at this time.

At one time in the past, I think the pygments project used pyparsing to crack Python code. They may have switched to another parser framework (or I may just have misremembered that this ever was the case).

It would be very helpful if you wrote a small parser and the corresponding syntax highlighting code for one of these environments, and attach both to this discussion thread. Or take one of the parsers in the pyparsing examples directory and write a syntax highlighter for it. I don't guarantee that I (or anyone) will pick up the project after you do this, but you might gain some insights for yourself in the process.
